EIPS/eip-signTypedData.md

```diff
@@ -0,0 +1,439 @@
+
+The above definition is not obviously collision free. Does a bytestring starting with `"x19Ethereum Signed Message:\n42a…"` mean a 42-byte string starting with `a` or a 4-byte string starting with `2a`?. This was pointed out in [Geth issue #14794][geth-issue-14794] and motivated Trezor to [not implement the standard][trezor] as-is. Fortunately this does not lead to actual collisions as the length of the encoded bytestring provides sufficient information to disambiguate the cases.
+
+[geth-issue-14794]: https://github.com/ethereum/go-ethereum/issues/14794
+[trezor]: https://github.com/trezor/trezor-mcu/issues/163
+
+Both determinism and invectiveness would be trivially true if `len(b)` was left out entirely. The point is, it is difficult to map arbitrary sets to bytestrings without introducing security issues in the encoding function. Yet the current design of `eth_sign` still takes a bytestring as input and expects implementors to come up with an encoding.
+
+### Messages
+
+The `eth_sign` call assumes messages to be bytestrings. In practice we are not hashing bytestrings but the collection of all semantically different messages of all different DApps `𝕄`. Unfortunately, this set is impossible to formalize  so we approximate it with the set of typed named structures `𝕊` and a domain separator `𝔹²⁵⁶` to obtain the set `𝔹²⁵⁶ × 𝕊`. This standard formalizes the set `𝕊` and provides a deterministic injective encoding function for `𝔹²⁵⁶ × 𝕊`.
+
+
+## Specification
+
+<!-- The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (cpp-ethereum, go-ethereum, parity, ethereumj, ethereumjs, ...).  -->
+
+The set of signable messages is extended from transactions and bytestrings `𝕋 ∪ 𝔹⁸ⁿ` to also include structured data `𝕊`. The new set of signable messages is thus `𝕋 ∪ 𝔹⁸ⁿ ∪ 𝕊`. They are encoded to bytestrings suitable for hashing and signing as follows:
+
+* `encode(t : 𝕋) = RLP_encode(t)`
+* `encode(b : 𝔹⁸ⁿ) = "\x19Ethereum Signed Message:\n" ‖ len(b) ‖ b` where `len(b)` is the *non-zero-padded* ascii-decimal encoding of the number of bytes in `b`.
+* `encode((d, s) : 𝔹²⁵⁶ × 𝕊) = "\x01" ‖ d ‖ hashStruct(s)` where `d` is a domain separator and `hashStruct(s)` is defined below.
+
+This encoding is deterministic because the individual components are. The encoding is injective because the three cases always differ in first byte. (`RLP_encode(t)` does not start with `\x01` or `\x19`.)
+
+### Domain separator `d`
+
+The domain separator is a 256-bit nonce. For each distinct use-case a new nonce is generated. It is recommended to use the address of a relevant contract or a randomly generated 256-bit constant. It should be considered part of the interface specification and shared with it.
```

Comment:
It seems like you're saying this should be the same across all instances of a given use-case. If so, calling this a nonce is very confusing, since nonces should only be used once.

Reply:
I agree. A 'nonce' is normally unique for a specific message, but here it is unique for a use-case, which is confusing. Other than that, the term 'nonce' is more accurate that say 'salt' or 'key'. It's a non-secret value that needs to be unique in a particular sense to prevent certain security issues. I'm not aware of a better term. Perhaps I should just call it 'domain separating number'.

I'll drop the term 'nonce' and think more carefully about the wording here. This part of the standard might also change a bit to incorporate @naure's ideas.